qiskit_experiments/curve_analysis/curve_analysis.py

```diff
@@ -274,39 +256,39 @@ def _format_data(
 
         columns = list(curve_data.columns)
         sort_by = itemgetter(
-            columns.index("model_id"),
+            columns.index("class_id"),
             columns.index("xval"),
         )
         # Use python native groupby method on ndarray. This is more performant than pandas one.
         average = averaging_methods[self.options.average_method]
+        model_names = self.model_names()
         formatted = []
-        for (mid, xv), g in groupby(sorted(curve_data.values, key=sort_by), key=sort_by):
+        for (class_id, xv), g in groupby(sorted(curve_data.values, key=sort_by), key=sort_by):
+            if class_id == 0:
+                # This is unclassified data
+                continue
             g_values = np.array(list(g))
             g_dict = dict(zip(columns, g_values.T))
             avg_yval, avg_yerr, shots = average(g_dict["yval"], g_dict["yerr"], g_dict["shots"])
+            data_name = g_dict["name"][0]
+            try:
+                # Map data index to model index through assigned name.
+                # Data name should match with the model name.
+                # Otherwise, the model index is unclassified.
+                model_id = model_names.index(data_name)
+            except ValueError:
+                model_id = pd.NA
             averaged = dict.fromkeys(columns)
+            averaged["category"] = category
             averaged["xval"] = xv
             averaged["yval"] = avg_yval
             averaged["yerr"] = avg_yerr
-            averaged["model_id"] = mid
+            averaged["name"] = g_dict["name"][0]
```

Comment:
Do you suggest removing name field if nothing matches? This could be a bit confusing as you say, but this requires Stark analysis to overwrite the entire method just to keep the data name for visualization of formatted data. Another solution would be re-adding separate `model_name` and `model_id` field, but for almost all analyses this information is duplicated and may confuse the users.